core/tests/unit/api_core/test_path_template.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2017 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import unicode_literals
+
+import mock
+import pytest
+
+from google.api.core import path_template
+
+
+@pytest.mark.parametrize('tmpl, args, kwargs, expected_result', [
+    # Basic positional params
+    ['/v1/*', ['a'], {}, '/v1/a'],
+    ['/v1/**', ['a/b'], {}, '/v1/a/b'],
+    ['/v1/*/*', ['a', 'b'], {}, '/v1/a/b'],
+    ['/v1/*/*/**', ['a', 'b', 'c/d'], {}, '/v1/a/b/c/d'],
+    # Basic named params
+    ['/v1/{name}', [], {'name': 'parent'}, '/v1/parent'],
+    ['/v1/{name=**}', [], {'name': 'parent/child'}, '/v1/parent/child'],
+    # Named params with a sub-template
+    ['/v1/{name=parent/*}', [], {'name': 'parent/child'}, '/v1/parent/child'],
+    ['/v1/{name=parent/**}', [], {'name': 'parent/child/object'},
+     '/v1/parent/child/object'],
+    # Combining positional and named params
+    ['/v1/*/{name}', ['a'], {'name': 'parent'}, '/v1/a/parent'],
+    ['/v1/{name}/*', ['a'], {'name': 'parent'}, '/v1/parent/a'],
+    ['/v1/{parent}/*/{child}/*', ['a', 'b'],
+     {'parent': 'parent', 'child': 'child'}, '/v1/parent/a/child/b'],
```

Comment:
Seems that having the template name and value being the same could be a source of bugs. Rename the values? E.g. `{'parent': 'zorp', 'child': 'zap'}`